## Nuclei Technologies Internship Project-1

## Predict the Sales of BigMart Stores

## Use Different Machine Learning Algorithms to Predict the Sales
## Evaluation Metric will be Root Mean Square Error

In [168]:
## Import the necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import mode

In [169]:
# Import data
train = pd.read_csv('D:\\Python\\Nuclei Technologies\\Internship Project\\Train.csv')
test = pd.read_csv('D:\\Python\\Nuclei Technologies\\Internship Project\\Test.csv')

### Checking basics in Data

In [170]:
train.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [171]:
test.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  \
0           FDW58       20.750          Low Fat         0.007565  Snack Foods   
1           FDW14        8.300              reg         0.038428        Dairy   
2           NCN55       14.600          Low Fat         0.099575       Others   
3           FDQ58        7.315          Low Fat         0.015388  Snack Foods   
4           FDY38          NaN          Regular         0.118599        Dairy   

   Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
0  107.8622            OUT049                       1999      Medium   
1   87.3198            OUT017                       2007         NaN   
2  241.7538            OUT010                       1998         NaN   
3  155.0340            OUT017                       2007         NaN   
4  234.2300            OUT027                       1985      Medium   

  Outlet_Location_Type        Outlet_Type  
0               Tier 1  Supermarket Type1  
1               Tier 2  Supermarket Type1  
2               Tier 3      Grocery Store  
3               Tier 2  Supermarket Type1  
4               Tier 3  Supermarket Type3

In [172]:
train.shape

(8523, 12)

In [173]:
test.shape

(5681, 11)

In [174]:
## Only difference is Item_Outlet_Sales column in Train data which is absent in Test data

In [175]:
##combining the datasets to work on both together, rather than to perform same steps on both separately
## marking 'train' and 'test' in dataset column
train['dataset']='train'
test['dataset']='test'

In [176]:
train.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales dataset  
0  Supermarket Type1          3735.1380   train  
1  Supermarket Type2           443.4228   train  
2  Supermarket Type1          2097.2700   train  
3      Grocery Store           732.3800   train  
4  Supermarket Type1           994.7052   train

In [177]:
test.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  \
0           FDW58       20.750          Low Fat         0.007565  Snack Foods   
1           FDW14        8.300              reg         0.038428        Dairy   
2           NCN55       14.600          Low Fat         0.099575       Others   
3           FDQ58        7.315          Low Fat         0.015388  Snack Foods   
4           FDY38          NaN          Regular         0.118599        Dairy   

   Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
0  107.8622            OUT049                       1999      Medium   
1   87.3198            OUT017                       2007         NaN   
2  241.7538            OUT010                       1998         NaN   
3  155.0340            OUT017                       2007         NaN   
4  234.2300            OUT027                       1985      Medium   

  Outlet_Location_Type        Outlet_Type dataset  
0               Tier 1  Supermarket Type1    test  
1               Tier 2  Supermarket Type1    test  
2               Tier 3      Grocery Store    test  
3               Tier 2  Supermarket Type1    test  
4               Tier 3  Supermarket Type3    test

In [178]:
## check the values
print(train.shape, test.shape)

(8523, 13) (5681, 12)


In [179]:
## Concatenating both datsets
data = pd.concat([train,test], ignore_index=True)

In [180]:
print(train.shape, test.shape, data.shape)

(8523, 13) (5681, 12) (14204, 13)


In [181]:
data.head(1)

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility Item_Type  \
0           FDA15          9.3          Low Fat         0.016047     Dairy   

   Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
0  249.8092            OUT049                       1999      Medium   

  Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales dataset  
0               Tier 1  Supermarket Type1           3735.138   train

In [182]:
data.tail(1)

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
14203           FDU37          9.5          Regular          0.10472   

      Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
14203    Canned    79.796            OUT045                       2002   

      Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  \
14203         NaN               Tier 2  Supermarket Type1                NaN   

      dataset  
14203    test

In [183]:
## Checking details of data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14204 entries, 0 to 14203
Data columns (total 13 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            14204 non-null  object 
 1   Item_Weight                11765 non-null  float64
 2   Item_Fat_Content           14204 non-null  object 
 3   Item_Visibility            14204 non-null  float64
 4   Item_Type                  14204 non-null  object 
 5   Item_MRP                   14204 non-null  float64
 6   Outlet_Identifier          14204 non-null  object 
 7   Outlet_Establishment_Year  14204 non-null  int64  
 8   Outlet_Size                10188 non-null  object 
 9   Outlet_Location_Type       14204 non-null  object 
 10  Outlet_Type                14204 non-null  object 
 11  Item_Outlet_Sales          8523 non-null   float64
 12  dataset                    14204 non-null  object 
dtypes: float64(4), int64(1), object(8)
memory usag

In [184]:
## checking Null values
data.isnull().sum()

Item_Identifier                 0
Item_Weight                  2439
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  4016
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales            5681
dataset                         0
dtype: int64

## Observations from above
1. Item_Outlet_Sales has 5681 Null values that is exactly the no. of data rows in Test so that is fine
2. Item_Weight and Outlet_Size will have to be treated

In [185]:
## Checking Numericals in data
data.describe()

Item_Weight  Item_Visibility      Item_MRP  Outlet_Establishment_Year  \
count  11765.000000     14204.000000  14204.000000               14204.000000   
mean      12.792854         0.065953    141.004977                1997.830681   
std        4.652502         0.051459     62.086938                   8.371664   
min        4.555000         0.000000     31.290000                1985.000000   
25%        8.710000         0.027036     94.012000                1987.000000   
50%       12.600000         0.054021    142.247000                1999.000000   
75%       16.750000         0.094037    185.855600                2004.000000   
max       21.350000         0.328391    266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800

## Observations from above 
1. Item_Visibility is zero for some items
2. Year of establishment ranges from 1985 to 2009

In [186]:
## Checking categorical data
data.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales dataset  
0  Supermarket Type1          3735.1380   train  
1  Supermarket Type2           443.4228   train  
2  Supermarket Type1          2097.2700   train  
3      Grocery Store           732.3800   train  
4  Supermarket Type1           994.7052   train

In [187]:
data['Item_Fat_Content'].value_counts()

Low Fat    8485
Regular    4824
LF          522
reg         195
low fat     178
Name: Item_Fat_Content, dtype: int64

## Observations from above
1. It seems that 'LF' and 'low fat' are same as 'Low Fat'
2. Also, 'reg' seems to be 'Regular' only
We will have to treat them accordingly.

In [188]:
data['Item_Type'].value_counts()

Fruits and Vegetables    2013
Snack Foods              1989
Household                1548
Frozen Foods             1426
Dairy                    1136
Baking Goods             1086
Canned                   1084
Health and Hygiene        858
Meat                      736
Soft Drinks               726
Breads                    416
Hard Drinks               362
Others                    280
Starchy Foods             269
Breakfast                 186
Seafood                    89
Name: Item_Type, dtype: int64

In [189]:
data['Outlet_Identifier'].value_counts()

OUT027    1559
OUT013    1553
OUT035    1550
OUT046    1550
OUT049    1550
OUT045    1548
OUT018    1546
OUT017    1543
OUT010     925
OUT019     880
Name: Outlet_Identifier, dtype: int64

## Observation only 10 unique outlets

In [190]:
data['Outlet_Size'].value_counts()

Medium    4655
Small     3980
High      1553
Name: Outlet_Size, dtype: int64

In [191]:
data['Outlet_Location_Type'].value_counts()

Tier 3    5583
Tier 2    4641
Tier 1    3980
Name: Outlet_Location_Type, dtype: int64

In [192]:
data['Outlet_Type'].value_counts()

Supermarket Type1    9294
Grocery Store        1805
Supermarket Type3    1559
Supermarket Type2    1546
Name: Outlet_Type, dtype: int64

In [193]:
## Imputing Null Values for Item_Weight
## Since its Numerical and Continuous we will replace Null values with its mean value
data['Item_Weight'].median()

12.6

In [194]:
## Imputing Null values in 'Item_Weight'
data['Item_Weight']=data['Item_Weight'].fillna(data['Item_Weight'].median())

In [195]:
## Check if Null Values have been imputed
data['Item_Weight'].isnull().sum()

0

In [196]:
##Check and Treat Null Values of Outlet_Size
## As it is Categorical we will impute Null values with its Mode value
data['Outlet_Size'].mode()

0    Medium
dtype: object

In [197]:
## Impute its Null Values
data['Outlet_Size']=data['Outlet_Size'].fillna('Medium')

In [198]:
data['Outlet_Size'].isnull().sum()

0

In [199]:
##Check overall Null Values
data.isnull().sum()

Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                     0
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales            5681
dataset                         0
dtype: int64

In [200]:
## Although Item_Visibility does not have Null Value but it can't have Zero value for certain items,
##as items will have some visibility however small it might be
## so replacing with its median value
data['Item_Visibility'].median()

0.05402053650000001

In [201]:
data['Item_Visibility']=data['Item_Visibility'].replace(to_replace=0, value=0.054020)

In [202]:
data['Item_Visibility'].value_counts()

0.054020    879
0.077169      3
0.076483      3
0.077290      3
0.076841      3
           ... 
0.064142      1
0.209684      1
0.019592      1
0.013530      1
0.115072      1
Name: Item_Visibility, Length: 13006, dtype: int64

In [203]:
## Replacing 'LF' and 'low fat' with 'Low Fat'
data['Item_Fat_Content']=data['Item_Fat_Content'].replace(to_replace='LF', value='Low Fat')
data['Item_Fat_Content']=data['Item_Fat_Content'].replace(to_replace='low fat', value='Low Fat')
## Replacing 'reg' with 'Regular'
data['Item_Fat_Content']=data['Item_Fat_Content'].replace(to_replace='reg', value='Regular')

In [204]:
data['Item_Fat_Content'].value_counts()

Low Fat    9185
Regular    5019
Name: Item_Fat_Content, dtype: int64

In [205]:
## Outlet Count
outlet_count = data['Outlet_Identifier'].value_counts()

In [206]:
outlet_count

OUT027    1559
OUT013    1553
OUT035    1550
OUT046    1550
OUT049    1550
OUT045    1548
OUT018    1546
OUT017    1543
OUT010     925
OUT019     880
Name: Outlet_Identifier, dtype: int64

In [207]:
## Item count
item_count = data['Item_Identifier'].value_counts()

In [208]:
item_count

FDR23    10
FDO11    10
NCT18    10
FDU26    10
DRF01    10
         ..
NCL42     7
FDL50     7
FDM50     7
FDO33     7
FDM10     7
Name: Item_Identifier, Length: 1559, dtype: int64

In [209]:
data['Outlet_Years']=2013 - data['Outlet_Establishment_Year']

In [210]:
data.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.054020   
4           NCD19         8.93          Low Fat         0.054020   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998      Medium               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales dataset  Outlet_Years  
0  Supermarket Type1          3735.1380   train            14  
1  Supermarket Type2           443.4228   train             4  
2  Supermarket Type1          2097.2700   train            14  
3      Grocery Store           732.3800   train            15  
4  Supermarket Type1           994.7052   train            26

In [211]:
## making a new column Item_Identifier_New for categorizing in three broad categories
data['Item_Identifier_New'] = data['Item_Identifier'].apply(lambda x : x[0:2])

In [212]:
data.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.054020   
4           NCD19         8.93          Low Fat         0.054020   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998      Medium               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales dataset  Outlet_Years  \
0  Supermarket Type1          3735.1380   train            14   
1  Supermarket Type2           443.4228   train             4   
2  Supermarket Type1          2097.2700   train            14   
3      Grocery Store           732.3800   train            15   
4  Supermarket Type1           994.7052   train            26   

  Item_Identifier_New  
0                  FD  
1                  DR  
2                  FD  
3                  FD  
4                  NC

In [213]:
## expanding the category to appropriate name identifier
data['Item_Identifier_New']=data['Item_Identifier_New'].map({'FD':'Food','NC':'Non-Consumable','DR':'Drinks'})

In [214]:
data.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.054020   
4           NCD19         8.93          Low Fat         0.054020   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998      Medium               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales dataset  Outlet_Years  \
0  Supermarket Type1          3735.1380   train            14   
1  Supermarket Type2           443.4228   train             4   
2  Supermarket Type1          2097.2700   train            14   
3      Grocery Store           732.3800   train            15   
4  Supermarket Type1           994.7052   train            26   

  Item_Identifier_New  
0                Food  
1              Drinks  
2                Food  
3                Food  
4      Non-Consumable

In [215]:
data.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dataset                       object
Outlet_Years                   int64
Item_Identifier_New           object
dtype: object

In [216]:
## Categorical Variables are 'Item_Fat_Content', 'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type', Item_Identifier_New

In [217]:
##Check Fat content where Items are Non-Consumable
data[data['Item_Identifier_New'] =='Non-Consumable']

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
4               NCD19         8.93          Low Fat         0.054020   
16              NCB42        11.80          Low Fat         0.008596   
22              NCB30        14.60          Low Fat         0.025698   
25              NCD06        13.00          Low Fat         0.099887   
31              NCS17        18.60          Low Fat         0.080829   
...               ...          ...              ...              ...   
14162           NCN26        10.85          Low Fat         0.028842   
14164           NCI43        19.85          Low Fat         0.026009   
14165           NCL17         7.39          Low Fat         0.067885   
14195           NCH42         6.86          Low Fat         0.036594   
14201           NCO17        10.00          Low Fat         0.073529   

                Item_Type  Item_MRP Outlet_Identifier  \
4               Household   53.8614            OUT013   
16     Health and Hygiene  115.3492            OUT018   
22              Household  196.5084            OUT035   
25              Household   45.9060            OUT017   
31     Health and Hygiene   96.4436            OUT018   
...                   ...       ...               ...   
14162           Household  116.9808            OUT017   
14164           Household   49.8376            OUT049   
14165  Health and Hygiene  143.0812            OUT049   
14195           Household  231.1010            OUT049   
14201  Health and Hygiene  118.7440            OUT045   

       Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
4                           1987        High               Tier 3   
16                          2009      Medium               Tier 3   
22                          2004       Small               Tier 2   
25                          2007      Medium               Tier 2   
31                          2009      Medium               Tier 3   
...                          ...         ...                  ...   
14162                       2007      Medium               Tier 2   
14164                       1999      Medium               Tier 1   
14165                       1999      Medium               Tier 1   
14195                       1999      Medium               Tier 1   
14201                       2002      Medium               Tier 2   

             Outlet_Type  Item_Outlet_Sales dataset  Outlet_Years  \
4      Supermarket Type1           994.7052   train            26   
16     Supermarket Type2          1621.8888   train             4   
22     Supermarket Type1          1587.2672   train             9   
25     Supermarket Type1           838.9080   train             6   
31     Supermarket Type2          2741.7644   train             4   
...                  ...                ...     ...           ...   
14162  Supermarket Type1                NaN    test             6   
14164  Supermarket Type1                NaN    test            14   
14165  Supermarket Type1                NaN    test            14   
14195  Supermarket Type1                NaN    test            14   
14201  Supermarket Type1                NaN    test            11   

      Item_Identifier_New  
4          Non-Consumable  
16         Non-Consumable  
22         Non-Consumable  
25         Non-Consumable  
31         Non-Consumable  
...                   ...  
14162      Non-Consumable  
14164      Non-Consumable  
14165      Non-Consumable  
14195      Non-Consumable  
14201      Non-Consumable  

[2686 rows x 15 columns]

In [218]:
##Replace the fat content of Non-Consumable Items as Non-Edible
data.loc[data['Item_Identifier_New'] =='Non-Consumable', 'Item_Fat_Content'] = "Non-Edible"

In [219]:
##Check again
data[data['Item_Identifier_New'] =='Non-Consumable']

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
4               NCD19         8.93       Non-Edible         0.054020   
16              NCB42        11.80       Non-Edible         0.008596   
22              NCB30        14.60       Non-Edible         0.025698   
25              NCD06        13.00       Non-Edible         0.099887   
31              NCS17        18.60       Non-Edible         0.080829   
...               ...          ...              ...              ...   
14162           NCN26        10.85       Non-Edible         0.028842   
14164           NCI43        19.85       Non-Edible         0.026009   
14165           NCL17         7.39       Non-Edible         0.067885   
14195           NCH42         6.86       Non-Edible         0.036594   
14201           NCO17        10.00       Non-Edible         0.073529   

                Item_Type  Item_MRP Outlet_Identifier  \
4               Household   53.8614            OUT013   
16     Health and Hygiene  115.3492            OUT018   
22              Household  196.5084            OUT035   
25              Household   45.9060            OUT017   
31     Health and Hygiene   96.4436            OUT018   
...                   ...       ...               ...   
14162           Household  116.9808            OUT017   
14164           Household   49.8376            OUT049   
14165  Health and Hygiene  143.0812            OUT049   
14195           Household  231.1010            OUT049   
14201  Health and Hygiene  118.7440            OUT045   

       Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
4                           1987        High               Tier 3   
16                          2009      Medium               Tier 3   
22                          2004       Small               Tier 2   
25                          2007      Medium               Tier 2   
31                          2009      Medium               Tier 3   
...                          ...         ...                  ...   
14162                       2007      Medium               Tier 2   
14164                       1999      Medium               Tier 1   
14165                       1999      Medium               Tier 1   
14195                       1999      Medium               Tier 1   
14201                       2002      Medium               Tier 2   

             Outlet_Type  Item_Outlet_Sales dataset  Outlet_Years  \
4      Supermarket Type1           994.7052   train            26   
16     Supermarket Type2          1621.8888   train             4   
22     Supermarket Type1          1587.2672   train             9   
25     Supermarket Type1           838.9080   train             6   
31     Supermarket Type2          2741.7644   train             4   
...                  ...                ...     ...           ...   
14162  Supermarket Type1                NaN    test             6   
14164  Supermarket Type1                NaN    test            14   
14165  Supermarket Type1                NaN    test            14   
14195  Supermarket Type1                NaN    test            14   
14201  Supermarket Type1                NaN    test            11   

      Item_Identifier_New  
4          Non-Consumable  
16         Non-Consumable  
22         Non-Consumable  
25         Non-Consumable  
31         Non-Consumable  
...                   ...  
14162      Non-Consumable  
14164      Non-Consumable  
14165      Non-Consumable  
14195      Non-Consumable  
14201      Non-Consumable  

[2686 rows x 15 columns]

In [220]:
#One Hot Coding:
data = pd.get_dummies(data, columns=['Item_Fat_Content', 'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type', 'Outlet_Identifier', 'Item_Identifier_New'])

In [221]:
data.head()

Item_Identifier  Item_Weight  Item_Visibility              Item_Type  \
0           FDA15         9.30         0.016047                  Dairy   
1           DRC01         5.92         0.019278            Soft Drinks   
2           FDN15        17.50         0.016760                   Meat   
3           FDX07        19.20         0.054020  Fruits and Vegetables   
4           NCD19         8.93         0.054020              Household   

   Item_MRP  Outlet_Establishment_Year  Item_Outlet_Sales dataset  \
0  249.8092                       1999          3735.1380   train   
1   48.2692                       2009           443.4228   train   
2  141.6180                       1999          2097.2700   train   
3  182.0950                       1998           732.3800   train   
4   53.8614                       1987           994.7052   train   

   Outlet_Years  Item_Fat_Content_Low Fat  ...  Outlet_Identifier_OUT018  \
0            14                         1  ...                         0   
1             4                         0  ...                         1   
2            14                         1  ...                         0   
3            15                         0  ...                         0   
4            26                         0  ...                         0   

   Outlet_Identifier_OUT019  Outlet_Identifier_OUT027  \
0                         0                         0   
1                         0                         0   
2                         0                         0   
3                         0                         0   
4                         0                         0   

   Outlet_Identifier_OUT035  Outlet_Identifier_OUT045  \
0                         0                         0   
1                         0                         0   
2                         0                         0   
3                         0                         0   
4                         0                         0   

   Outlet_Identifier_OUT046  Outlet_Identifier_OUT049  \
0                         0                         1   
1                         0                         0   
2                         0                         1   
3                         0                         0   
4                         0                         0   

   Item_Identifier_New_Drinks  Item_Identifier_New_Food  \
0                           0                         1   
1                           1                         0   
2                           0                         1   
3                           0                         1   
4                           0                         0   

   Item_Identifier_New_Non-Consumable  
0                                   0  
1                                   0  
2                                   0  
3                                   0  
4                                   1  

[5 rows x 35 columns]

In [222]:
data.dtypes

Item_Identifier                        object
Item_Weight                           float64
Item_Visibility                       float64
Item_Type                              object
Item_MRP                              float64
Outlet_Establishment_Year               int64
Item_Outlet_Sales                     float64
dataset                                object
Outlet_Years                            int64
Item_Fat_Content_Low Fat                uint8
Item_Fat_Content_Non-Edible             uint8
Item_Fat_Content_Regular                uint8
Outlet_Size_High                        uint8
Outlet_Size_Medium                      uint8
Outlet_Size_Small                       uint8
Outlet_Location_Type_Tier 1             uint8
Outlet_Location_Type_Tier 2             uint8
Outlet_Location_Type_Tier 3             uint8
Outlet_Type_Grocery Store               uint8
Outlet_Type_Supermarket Type1           uint8
Outlet_Type_Supermarket Type2           uint8
Outlet_Type_Supermarket Type3     

In [223]:
data.drop(['Item_Type','Outlet_Establishment_Year'],axis=1,inplace=True)

In [224]:
train1 = data.loc[data['dataset']=="train"]
test1 = data.loc[data['dataset']=="test"]

In [225]:
test1.drop(['Item_Identifier','Item_Outlet_Sales','dataset'],axis=1,inplace=True)
train1.drop(['Item_Identifier','dataset'],axis=1,inplace=True)

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [226]:
test1.head()

Item_Weight  Item_Visibility  Item_MRP  Outlet_Years  \
8523       20.750         0.007565  107.8622            14   
8524        8.300         0.038428   87.3198             6   
8525       14.600         0.099575  241.7538            15   
8526        7.315         0.015388  155.0340             6   
8527       12.600         0.118599  234.2300            28   

      Item_Fat_Content_Low Fat  Item_Fat_Content_Non-Edible  \
8523                         1                            0   
8524                         0                            0   
8525                         0                            1   
8526                         1                            0   
8527                         0                            0   

      Item_Fat_Content_Regular  Outlet_Size_High  Outlet_Size_Medium  \
8523                         0                 0                   1   
8524                         1                 0                   1   
8525                         0                 0                   1   
8526                         0                 0                   1   
8527                         1                 0                   1   

      Outlet_Size_Small  ...  Outlet_Identifier_OUT018  \
8523                  0  ...                         0   
8524                  0  ...                         0   
8525                  0  ...                         0   
8526                  0  ...                         0   
8527                  0  ...                         0   

      Outlet_Identifier_OUT019  Outlet_Identifier_OUT027  \
8523                         0                         0   
8524                         0                         0   
8525                         0                         0   
8526                         0                         0   
8527                         0                         1   

      Outlet_Identifier_OUT035  Outlet_Identifier_OUT045  \
8523                         0                         0   
8524                         0                         0   
8525                         0                         0   
8526                         0                         0   
8527                         0                         0   

      Outlet_Identifier_OUT046  Outlet_Identifier_OUT049  \
8523                         0                         1   
8524                         0                         0   
8525                         0                         0   
8526                         0                         0   
8527                         0                         0   

      Item_Identifier_New_Drinks  Item_Identifier_New_Food  \
8523                           0                         1   
8524                           0                         1   
8525                           0                         0   
8526                           0                         1   
8527                           0                         1   

      Item_Identifier_New_Non-Consumable  
8523                                   0  
8524                                   0  
8525                                   1  
8526                                   0  
8527                                   0  

[5 rows x 30 columns]

In [227]:
test1.columns

Index(['Item_Weight', 'Item_Visibility', 'Item_MRP', 'Outlet_Years',
       'Item_Fat_Content_Low Fat', 'Item_Fat_Content_Non-Edible',
       'Item_Fat_Content_Regular', 'Outlet_Size_High', 'Outlet_Size_Medium',
       'Outlet_Size_Small', 'Outlet_Location_Type_Tier 1',
       'Outlet_Location_Type_Tier 2', 'Outlet_Location_Type_Tier 3',
       'Outlet_Type_Grocery Store', 'Outlet_Type_Supermarket Type1',
       'Outlet_Type_Supermarket Type2', 'Outlet_Type_Supermarket Type3',
       'Outlet_Identifier_OUT010', 'Outlet_Identifier_OUT013',
       'Outlet_Identifier_OUT017', 'Outlet_Identifier_OUT018',
       'Outlet_Identifier_OUT019', 'Outlet_Identifier_OUT027',
       'Outlet_Identifier_OUT035', 'Outlet_Identifier_OUT045',
       'Outlet_Identifier_OUT046', 'Outlet_Identifier_OUT049',
       'Item_Identifier_New_Drinks', 'Item_Identifier_New_Food',
       'Item_Identifier_New_Non-Consumable'],
      dtype='object')

In [228]:
train1.head()

Item_Weight  Item_Visibility  Item_MRP  Item_Outlet_Sales  Outlet_Years  \
0         9.30         0.016047  249.8092          3735.1380            14   
1         5.92         0.019278   48.2692           443.4228             4   
2        17.50         0.016760  141.6180          2097.2700            14   
3        19.20         0.054020  182.0950           732.3800            15   
4         8.93         0.054020   53.8614           994.7052            26   

   Item_Fat_Content_Low Fat  Item_Fat_Content_Non-Edible  \
0                         1                            0   
1                         0                            0   
2                         1                            0   
3                         0                            0   
4                         0                            1   

   Item_Fat_Content_Regular  Outlet_Size_High  Outlet_Size_Medium  ...  \
0                         0                 0                   1  ...   
1                         1                 0                   1  ...   
2                         0                 0                   1  ...   
3                         1                 0                   1  ...   
4                         0                 1                   0  ...   

   Outlet_Identifier_OUT018  Outlet_Identifier_OUT019  \
0                         0                         0   
1                         1                         0   
2                         0                         0   
3                         0                         0   
4                         0                         0   

   Outlet_Identifier_OUT027  Outlet_Identifier_OUT035  \
0                         0                         0   
1                         0                         0   
2                         0                         0   
3                         0                         0   
4                         0                         0   

   Outlet_Identifier_OUT045  Outlet_Identifier_OUT046  \
0                         0                         0   
1                         0                         0   
2                         0                         0   
3                         0                         0   
4                         0                         0   

   Outlet_Identifier_OUT049  Item_Identifier_New_Drinks  \
0                         1                           0   
1                         0                           1   
2                         1                           0   
3                         0                           0   
4                         0                           0   

   Item_Identifier_New_Food  Item_Identifier_New_Non-Consumable  
0                         1                                   0  
1                         0                                   0  
2                         1                                   0  
3                         1                                   0  
4                         0                                   1  

[5 rows x 31 columns]

In [229]:
train1.columns

Index(['Item_Weight', 'Item_Visibility', 'Item_MRP', 'Item_Outlet_Sales',
       'Outlet_Years', 'Item_Fat_Content_Low Fat',
       'Item_Fat_Content_Non-Edible', 'Item_Fat_Content_Regular',
       'Outlet_Size_High', 'Outlet_Size_Medium', 'Outlet_Size_Small',
       'Outlet_Location_Type_Tier 1', 'Outlet_Location_Type_Tier 2',
       'Outlet_Location_Type_Tier 3', 'Outlet_Type_Grocery Store',
       'Outlet_Type_Supermarket Type1', 'Outlet_Type_Supermarket Type2',
       'Outlet_Type_Supermarket Type3', 'Outlet_Identifier_OUT010',
       'Outlet_Identifier_OUT013', 'Outlet_Identifier_OUT017',
       'Outlet_Identifier_OUT018', 'Outlet_Identifier_OUT019',
       'Outlet_Identifier_OUT027', 'Outlet_Identifier_OUT035',
       'Outlet_Identifier_OUT045', 'Outlet_Identifier_OUT046',
       'Outlet_Identifier_OUT049', 'Item_Identifier_New_Drinks',
       'Item_Identifier_New_Food', 'Item_Identifier_New_Non-Consumable'],
      dtype='object')

In [230]:
train1.to_csv("D:\\Python\\Nuclei Technologies\\Internship Project\\train_t.csv",index=False)
test1.to_csv("D:\\Python\\Nuclei Technologies\\Internship Project\\test_t.csv",index=False)

In [231]:
from sklearn.model_selection import train_test_split

In [232]:
y = train1['Item_Outlet_Sales']
X = train1.drop(columns=['Item_Outlet_Sales'],axis=1)

In [233]:
train1.head()

Item_Weight  Item_Visibility  Item_MRP  Item_Outlet_Sales  Outlet_Years  \
0         9.30         0.016047  249.8092          3735.1380            14   
1         5.92         0.019278   48.2692           443.4228             4   
2        17.50         0.016760  141.6180          2097.2700            14   
3        19.20         0.054020  182.0950           732.3800            15   
4         8.93         0.054020   53.8614           994.7052            26   

   Item_Fat_Content_Low Fat  Item_Fat_Content_Non-Edible  \
0                         1                            0   
1                         0                            0   
2                         1                            0   
3                         0                            0   
4                         0                            1   

   Item_Fat_Content_Regular  Outlet_Size_High  Outlet_Size_Medium  ...  \
0                         0                 0                   1  ...   
1                         1                 0                   1  ...   
2                         0                 0                   1  ...   
3                         1                 0                   1  ...   
4                         0                 1                   0  ...   

   Outlet_Identifier_OUT018  Outlet_Identifier_OUT019  \
0                         0                         0   
1                         1                         0   
2                         0                         0   
3                         0                         0   
4                         0                         0   

   Outlet_Identifier_OUT027  Outlet_Identifier_OUT035  \
0                         0                         0   
1                         0                         0   
2                         0                         0   
3                         0                         0   
4                         0                         0   

   Outlet_Identifier_OUT045  Outlet_Identifier_OUT046  \
0                         0                         0   
1                         0                         0   
2                         0                         0   
3                         0                         0   
4                         0                         0   

   Outlet_Identifier_OUT049  Item_Identifier_New_Drinks  \
0                         1                           0   
1                         0                           1   
2                         1                           0   
3                         0                           0   
4                         0                           0   

   Item_Identifier_New_Food  Item_Identifier_New_Non-Consumable  
0                         1                                   0  
1                         0                                   0  
2                         1                                   0  
3                         1                                   0  
4                         0                                   1  

[5 rows x 31 columns]

In [234]:
X.columns

Index(['Item_Weight', 'Item_Visibility', 'Item_MRP', 'Outlet_Years',
       'Item_Fat_Content_Low Fat', 'Item_Fat_Content_Non-Edible',
       'Item_Fat_Content_Regular', 'Outlet_Size_High', 'Outlet_Size_Medium',
       'Outlet_Size_Small', 'Outlet_Location_Type_Tier 1',
       'Outlet_Location_Type_Tier 2', 'Outlet_Location_Type_Tier 3',
       'Outlet_Type_Grocery Store', 'Outlet_Type_Supermarket Type1',
       'Outlet_Type_Supermarket Type2', 'Outlet_Type_Supermarket Type3',
       'Outlet_Identifier_OUT010', 'Outlet_Identifier_OUT013',
       'Outlet_Identifier_OUT017', 'Outlet_Identifier_OUT018',
       'Outlet_Identifier_OUT019', 'Outlet_Identifier_OUT027',
       'Outlet_Identifier_OUT035', 'Outlet_Identifier_OUT045',
       'Outlet_Identifier_OUT046', 'Outlet_Identifier_OUT049',
       'Item_Identifier_New_Drinks', 'Item_Identifier_New_Food',
       'Item_Identifier_New_Non-Consumable'],
      dtype='object')

In [235]:
y

0       3735.1380
1        443.4228
2       2097.2700
3        732.3800
4        994.7052
          ...    
8518    2778.3834
8519     549.2850
8520    1193.1136
8521    1845.5976
8522     765.6700
Name: Item_Outlet_Sales, Length: 8523, dtype: float64

## Linear Regression Model

In [236]:
## Referred https://www.analyticsvidhya.com/blog/2016/02/bigmart-sales-solution-top-20/

In [237]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=101)

In [238]:
from sklearn.linear_model import LinearRegression

In [239]:
lm = LinearRegression()

In [240]:
lm.fit(X_train, y_train)

LinearRegression()

In [241]:
y_pred=lm.predict(X_test)

In [82]:
from sklearn import metrics

## Linear Regression Metrics

In [83]:
print('MAE', metrics.mean_absolute_error(y_test,y_pred))
print('MSE', metrics.mean_squared_error(y_test,y_pred))
print('RMSE', np.sqrt(metrics.mean_squared_error(y_test,y_pred)))
print('R2 score', metrics.r2_score(y_test,y_pred))

MAE 809.0304310565749
MSE 1158515.3278059007
RMSE 1076.3434989843627
R2 score 0.568782337279371


## Decision Trees Model

In [84]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=101)

In [85]:
from sklearn.tree import DecisionTreeRegressor

In [86]:
dtree = DecisionTreeRegressor()

In [87]:
dtree.fit(X_train, y_train)

DecisionTreeRegressor()

In [88]:
y_pred = dtree.predict(X_test)

## Decision Tree Metrics

In [89]:
print('MAE', metrics.mean_absolute_error(y_test,y_pred))
print('MSE', metrics.mean_squared_error(y_test,y_pred))
print('RMSE', np.sqrt(metrics.mean_squared_error(y_test,y_pred)))
print('R2 score', metrics.r2_score(y_test,y_pred))

MAE 1066.6897149784904
MSE 2323450.3305208688
RMSE 1524.2868268540763
R2 score 0.1351751703862274


## Random Forest Model

In [90]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=101)

In [91]:
from sklearn.ensemble import RandomForestRegressor

In [92]:
rfc = RandomForestRegressor(n_estimators=300)

In [93]:
rfc.fit(X_train,y_train)

RandomForestRegressor(n_estimators=300)

In [94]:
y_pred = rfc.predict(X_test)

## Random Forest Metrics

In [95]:
print('MAE', metrics.mean_absolute_error(y_test,y_pred))
print('MSE', metrics.mean_squared_error(y_test,y_pred))
print('RMSE', np.sqrt(metrics.mean_squared_error(y_test,y_pred)))
print('R2 score', metrics.r2_score(y_test,y_pred))

MAE 764.3112394219788
MSE 1173164.1268761496
RMSE 1083.1270132704426
R2 score 0.5633298234065505


## XGBoost Model

In [96]:
## Referring https://machinelearningmastery.com/xgboost-for-regression/

In [97]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=101)

In [98]:
from xgboost import XGBRegressor

In [99]:
xgb = XGBRegressor(n_estimators=1000, max_depth=7, eta=0.1, subsample=0.7, colsample_bytree=0.8)

In [100]:
xgb.fit(X_train,y_train)

XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=0.8, eta=0.1, gamma=0,
             gpu_id=-1, importance_type='gain', interaction_constraints='',
             learning_rate=0.100000001, max_delta_step=0, max_depth=7,
             min_child_weight=1, missing=nan, monotone_constraints='()',
             n_estimators=1000, n_jobs=8, num_parallel_tree=1, random_state=0,
             reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=0.7,
             tree_method='exact', validate_parameters=1, verbosity=None)

In [101]:
y_pred = xgb.predict(X_test)

## XGBoost Metrics

In [102]:
print('MAE', metrics.mean_absolute_error(y_test,y_pred))
print('MSE', metrics.mean_squared_error(y_test,y_pred))
print('RMSE', np.sqrt(metrics.mean_squared_error(y_test,y_pred)))
print('R2 score', metrics.r2_score(y_test,y_pred))

MAE 844.2126424974464
MSE 1386836.9975752516
RMSE 1177.6404364555642
R2 score 0.4837974135383307


## In our case Linear Regression is performing best, so we will use the same to do prediction of 'Item_Outlet_Sales'


In [103]:
test1.head()

Item_Weight  Item_Visibility  Item_MRP  Outlet_Years  \
8523       20.750         0.007565  107.8622            14   
8524        8.300         0.038428   87.3198             6   
8525       14.600         0.099575  241.7538            15   
8526        7.315         0.015388  155.0340             6   
8527       12.600         0.118599  234.2300            28   

      Item_Fat_Content_Low Fat  Item_Fat_Content_Non-Edible  \
8523                         1                            0   
8524                         0                            0   
8525                         0                            1   
8526                         1                            0   
8527                         0                            0   

      Item_Fat_Content_Regular  Outlet_Size_High  Outlet_Size_Medium  \
8523                         0                 0                   1   
8524                         1                 0                   1   
8525                         0                 0                   1   
8526                         0                 0                   1   
8527                         1                 0                   1   

      Outlet_Size_Small  ...  Outlet_Identifier_OUT018  \
8523                  0  ...                         0   
8524                  0  ...                         0   
8525                  0  ...                         0   
8526                  0  ...                         0   
8527                  0  ...                         0   

      Outlet_Identifier_OUT019  Outlet_Identifier_OUT027  \
8523                         0                         0   
8524                         0                         0   
8525                         0                         0   
8526                         0                         0   
8527                         0                         1   

      Outlet_Identifier_OUT035  Outlet_Identifier_OUT045  \
8523                         0                         0   
8524                         0                         0   
8525                         0                         0   
8526                         0                         0   
8527                         0                         0   

      Outlet_Identifier_OUT046  Outlet_Identifier_OUT049  \
8523                         0                         1   
8524                         0                         0   
8525                         0                         0   
8526                         0                         0   
8527                         0                         0   

      Item_Identifier_New_Drinks  Item_Identifier_New_Food  \
8523                           0                         1   
8524                           0                         1   
8525                           0                         0   
8526                           0                         1   
8527                           0                         1   

      Item_Identifier_New_Non-Consumable  
8523                                   0  
8524                                   0  
8525                                   1  
8526                                   0  
8527                                   0  

[5 rows x 30 columns]

In [104]:
test1.shape

(5681, 30)

In [242]:
y_pred=lm.predict(test1)

In [243]:
print(y_pred)

[1877.79886283 1597.11425763 1874.02062782 ... 1807.68521722 3582.76165609
 1253.28906871]


In [244]:
type(y_pred)

numpy.ndarray

In [245]:
y_pred = y_pred.astype(int)

In [246]:
type(y_pred)

numpy.ndarray

In [247]:
df2 = pd.DataFrame(y_pred,columns=['Item_Outlet_Sales'])
df2.head()

Item_Outlet_Sales
0               1877
1               1597
2               1874
3               2630
4               5203

In [248]:
df2.to_csv("D:\\Python\\Nuclei Technologies\\Internship Project\\Final_Prediction.csv",index=False)